src/cdi_ndi/NDI_TN.cc

```diff
@@ -0,0 +1,279 @@
+//----------------------------------*-C++-*-----------------------------------//
+/*!
+ * \file   cdi_ndi/NDI_TN.cc
+ * \author Ben R. Ryan
+ * \date   2020 Feb 4
+ * \brief  NDI_TN member definitions.
+ * \note   Copyright (C) 2019-2020 Triad National Security, LLC.
+ *         All rights reserved. */
+//----------------------------------------------------------------------------//
+
+#include "NDI_TN.hh"
+#include <cmath>
+
+namespace rtt_cdi_ndi {
+
+//----------------------------------------------------------------------------//
+// CONSTRUCTORS
+//----------------------------------------------------------------------------//
+/*!
+ * \brief Constructor for NDI reader specific to TN reaction data.
+ *
+ * This constructor opens an NDI file, navigates to the appropriate data, reads
+ * the data into internal buffers, and closes the file. For more details on NDI,
+ * see https://xweb.lanl.gov/projects/data/nuclear/ndi/ndi.html
+ *
+ * \param[in] gendir_in path to gendir file
+ * \param[in] library_in name of requested NDI data library
+ * \param[in] reaction_in name of requested reaction
+ * \param[in] discretization_in type of energy discretization to request
+ */
+NDI_TN::NDI_TN(const std::string &gendir_in, const std::string &library_in,
+               const std::string &reaction_in,
+               const DISCRETIZATION discretization_in)
+    : NDI_Base(gendir_in, "tn", library_in, reaction_in, discretization_in),
+      control(1234567) {
+  int gendir_handle, dataset_handle, ndi_size_error;
+  constexpr int c_str_len = 4096;
+  char c_str_buf[c_str_len];
+
+  control.initialize(1, rng);
+
+  // Open gendir file (index of a complete NDI dataset)
+  SAFE_NDI(NDI2_open_gendir(&gendir_handle, gendir.c_str()));
+
+  // Set dataset option
+  SAFE_NDI(NDI2_set_option_gendir(gendir_handle, NDI_LIB_TYPE_DEFAULT,
+                                  dataset.c_str()));
+
+  //! Set library option
+  SAFE_NDI(NDI2_set_option_gendir(gendir_handle, NDI_LIBRARY_DEFAULT,
+                                  library.c_str()));
+
```

Comment:
This checking is best left to NDI, I think, since we won't have knowledge of the available libraries except through NDI. If the library is not valid, this call will return `ndi_error != 0` which we will catch immediately afterwards.